File: typescript/openai-raw.ts
/**
 * Calls the OpenAI Chat Completions API via raw fetch (no SDK).
 * Showcases model usage through direct HTTP.
 */

const OPENAI_API_HOST = "https://api.openai.com";
const OPENAI_CHAT_PATH = "/v1/chat/completions";
const OPENAI_MODEL = "gpt-4o-mini";

async function getChatResponse(promptMessage: string): Promise<string> {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey || apiKey === "YOUR_OPENAI_API_KEY") {
    return "Error: Set OPENAI_API_KEY in the environment.";
  }

  try {
    const response = await fetch(`${OPENAI_API_HOST}${OPENAI_CHAT_PATH}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model: OPENAI_MODEL,
        messages: [{ role: "user" as const, content: promptMessage }],
        temperature: 0.7,
        max_tokens: 150,
      }),
    });

    if (!response.ok) {
      const text = await response.text();
      return `API Error: ${response.status} - ${response.statusText}. ${text}`;
    }

    const data = (await response.json()) as {
      choices?: Array<{ message?: { content?: string } }>;
    };
    const content = data.choices?.[0]?.message?.content;
    return content ?? "No response from chatbot.";
  } catch (err) {
    return `Error: ${err instanceof Error ? err.message : String(err)}`;
  }
}

async function main(): Promise<void> {
  console.log("OpenAI Chat (raw fetch). Type 'exit' to quit.\n");

  const readline = await import("readline");
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

  const ask = (): void => {
    rl.question("You: ", async (userInput) => {
      if (userInput.trim().toLowerCase() === "exit") {
        console.log("Goodbye!");
        rl.close();
        return;
      }
      console.log("Chatbot: Thinking...");
      const reply = await getChatResponse(userInput);
      console.log(`Chatbot: ${reply}\n`);
      ask();
    });
  };

  ask();
}

main();

File: java/src/main/java/aibuffet/OpenAiRaw.java
package aibuffet;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Scanner;

/**
 * Calls the OpenAI Chat Completions API via raw HTTP (no SDK).
 * Showcases model usage through direct HTTP.
 */
public final class OpenAiRaw {

    private static final String OPENAI_API_HOST = "https://api.openai.com";
    private static final String OPENAI_CHAT_PATH = "/v1/chat/completions";
    private static final String OPENAI_MODEL = "gpt-4o-mini";

    private static String getChatResponse(String promptMessage) {
        String apiKey = System.getenv("OPENAI_API_KEY");
        if (apiKey == null || apiKey.isBlank() || "YOUR_OPENAI_API_KEY".equals(apiKey)) {
            return "Error: Set OPENAI_API_KEY in the environment.";
        }

        try {
            String body = "{"
                    + "\"model\": \"" + OPENAI_MODEL + "\","
                    + "\"messages\": [{\"role\": \"user\", \"content\": " + jsonString(promptMessage) + "}],"
                    + "\"temperature\": 0.7,"
                    + "\"max_tokens\": 150"
                    + "}";

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(OPENAI_API_HOST + OPENAI_CHAT_PATH))
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer " + apiKey)
                    .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                    .build();

            HttpResponse<String> response = HttpClient.newHttpClient()
                    .send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() >= 400) {
                return "API Error: " + response.statusCode() + " - " + response.body();
            }

            return extractContent(response.body());
        } catch (Exception err) {
            return "Error: " + err.getMessage();
        }
    }

    private static String jsonString(String value) {
        return "\"" + value.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
    }

    private static String extractContent(String json) {
        String marker = "\"content\":\"";
        int start = json.indexOf(marker);
        if (start < 0) {
            return "No response from chatbot.";
        }
        start += marker.length();
        StringBuilder content = new StringBuilder();
        for (int i = start; i < json.length(); i++) {
            char c = json.charAt(i);
            if (c == '\\' && i + 1 < json.length()) {
                content.append(json.charAt(i + 1));
                i++;
                continue;
            }
            if (c == '"') {
                break;
            }
            content.append(c);
        }
        return content.length() == 0 ? "No response from chatbot." : content.toString();
    }

    public static void main(String[] args) {
        System.out.println("OpenAI Chat (raw HTTP). Type 'exit' to quit.\n");

        Scanner scanner = new Scanner(new BufferedReader(new InputStreamReader(System.in)));
        while (true) {
            System.out.print("You: ");
            if (!scanner.hasNextLine()) {
                break;
            }
            String userInput = scanner.nextLine();
            if ("exit".equalsIgnoreCase(userInput.trim())) {
                System.out.println("Goodbye!");
                break;
            }

            System.out.println("Chatbot: Thinking...");
            String reply = getChatResponse(userInput);
            System.out.println("Chatbot: " + reply + "\n");
        }
    }
}
